### ISSUE TITLE
"import matplotlib.pyplot" fails on NetBSD with version 2.2.2

### ISSUE THREAD

**Issue by gson1703** [matplotlib/matplotlib#11635]
### Bug report

As originally reported in http://gnats.netbsd.org/53425 , with matplotlib version 2.2.2, attempting to `import matplotlib.pyplot` on a NetBSD system fails with:

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/pkg/lib/python2.7/site-packages/matplotlib/pyplot.py", line 31, in <module>
    import matplotlib.colorbar
  File "/usr/pkg/lib/python2.7/site-packages/matplotlib/colorbar.py", line 36, in <module>
    import matplotlib.contour as contour
  File "/usr/pkg/lib/python2.7/site-packages/matplotlib/contour.py", line 20, in <module>
    import matplotlib.font_manager as font_manager
  File "/usr/pkg/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1469, in <module>
    _rebuild()
  File "/usr/pkg/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1450, in _rebuild
    fontManager = FontManager()
  File "/usr/pkg/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1086, in __init__
    self.ttflist = createFontList(self.ttffiles)
  File "/usr/pkg/lib/python2.7/site-packages/matplotlib/font_manager.py", line 587, in createFontList
    font = ft2font.FT2Font(fpath)
TypeError: First argument must be a path or file object reading bytes
```

I tracked down the cause of this to `mpl_PyFile_Dup()` attempting to call the "flush" method of a file object that is open only for reading.  Python implements this by calling `fflush()`, and calling `fflush()` on a file descriptor not open for writing is treated as an error on BSD systems (but silently ignored on Linux). The `fflush()` call returns -1 with errno=EBADF, and this ultimately gets translated into the rather unhelpful error message "TypeError: First argument must be a path or file object reading bytes" show above.

The call chain leading to the error is PyFT2Font_init() -> convert_open_args() -> mpl_PyFile_Dup().

**Code for reproduction**

On a BSD system, run

python2.7 -c 'import matplotlib.pyplot'

**Actual outcome**

The stack trace shown above is printed.

**Expected outcome**

The command silently exits with a success status.

**Matplotlib version**

  * Operating system: NetBSD 7.1_STABLE
  * Matplotlib version:  2.2.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): GTKAgg
  * Python version:  2.7.15 
  * Jupyter version (if applicable): N/A
  * Other libraries: N/A

matplotlib was installed via pkgsrc.

I have a patch which I will attempt to submit as a pull request.

**Comment by tacaswell** [comment]
@gson1703 Sorry to ping you again, but do you know what changed to make this an issue?  We have not changed that c code from ~2013 / 2014 and have not previously had issues on macOS or *BSD flavors.

**Comment by tacaswell** [comment]
This is likely not due to a change in cpython as the code around `file_flush` has not really been changed from 1992 (!!)

The freebsd fflush has not changed from 1993 (https://www.freebsd.org/cgi/man.cgi?query=fflush&sektion=3&apropos=0&manpath=freebsd).

In the standard state of a programmer looking at a bug "I have no idea why this ever worked...."

**Comment by gson1703** [comment]
@tacaswell  I believe the what changed was the following diff between matplotlib 2.1.2 and 2.2.2:

```
diff -ru matplotlib-2.1.2/src/file_compat.h matplotlib-2.2.2/src/file_compat.h
--- matplotlib-2.1.2/src/file_compat.h  2018-01-18 03:36:16.000000000 +0200
+++ matplotlib-2.2.2/src/file_compat.h  2018-03-06 06:47:09.000000000 +0200
@@ -48,7 +48,7 @@
 /*
  * PyFile_* compatibility
  */
-#if PY3K
+#if defined(PY3K) | defined(PYPY_VERSION)
```

The 2.1.2 version of the #if expression is false, so the following trivial version of `mpl_PyFile_Dup()` in the #else branch is used; it does not flush, and therefore does not trigger the bug:

```
static NPY_INLINE FILE *mpl_PyFile_Dup(PyObject *file, const char *mode, mpl_off_t *orig_pos)
{
    return PyFile_AsFile(file);
}
```

The 2.2.2 version of the #if expression is true, selecting the other, nontrivial version of `mpl_PyFile_Dup()`, which does a flush and therefore triggers the bug.

**Comment by gson1703** [comment]
@tacaswell That #if condition makes little sense because `defined(PY3K)` is always true. Whoever made that change probably meant  `#if PY3K || defined(PYPY_VERSION)`.

**Comment by tacaswell** [comment]
So that change came in via 68492fbd6926cbb55dae4dc12519a16382eaa8d7 so the entire 2.2.x series is nominally affected by this, but if we go back to 2 months ago (before @gson1703 's PR fixed things) travis was passing on 2.7 on OSX (ex https://travis-ci.org/matplotlib/matplotlib/jobs/395614611 ).

It seems for python3 all of the file I/O code did get re-written, now the only use of `fflush` in cpython is on stdout/stderr.
